### DISCUSSION TITLE
Temporary workaround for debugging Transmission with Xcode on macOS Sonoma

### DISCUSSION THREAD

**Discussion by Coeur** [transmission/transmission#5839] [Show and tell]
So, you've updated to macOS Sonoma 14.0 beta? Great.
Now you want to "debug" Transmission with Xcode 15 beta? Not so great. Your main UI is all grayed out while debugging.

I couldn't figure out the cause of this visual glitch (which happens even with very old source code of Transmission), but I found two workarounds: building with Xcode 14's `xcodebuild` and debugging with `Xcode 15`, or simply using `Xcode 14` IDE.

So first, get Xcode 14.3.1 (from App Store or from https://developer.apple.com/download/all/?q=xcode%2014) on your macOS Sonoma.

### Workaround with `xcodebuild`
Here are the steps of this workaround:
1. `sudo xcode-select -p /Applications/Xcode14.app/Contents/Developer`. So that your command line tools are those from Xcode 14.
2. `rm -rf build`. Assuming all commands are from the "transmission" root folder, we clean the build folder.
3. `cmake -B build -G Xcode -DCMAKE_BUILD_TYPE=RelWithDebInfo`. We make the project for Xcode.
4. `open build/transmission.xcodeproj`. We open the project with Xcode 15.
5. Go to the Project settings and change the Base SDK (SDKROOT) to "macOS" or "No SDK" (but don't choose "macOS 13.3" or "macOS 14.0" explicitly, that wouldn't compile).
6. Select "transmission-mac" as your scheme, and "Edit Scheme..." > Run > Launch > "Wait for the executable to be launched".
7. Start the active scheme (the "play" button -- @GaryElshaw 🤣) and let it build until you see "Waiting to attach to Transmission".
8. `xcodebuild -project build/transmission.xcodeproj -target transmission-mac`. We literally overwrite the previous `Xcode 15` build with an `xcodebuild` build from Xcode 14 command line tools.
9. `open ./build/macosx/Debug/Transmission.app`. We launch Transmission. Xcode 15 will attach itself to it. You can now use breakpoints in the app.

### Workaround with `Xcode 14` IDE
Here are the steps of this workaround:
1. Open Xcode 14 using the terminal command `/Applications/Xcode.app/Contents/MacOS/Xcode` or `open /Applications/Xcode.app/Contents/MacOS/Xcode`.

**Comment by nevack** [comment]
Wow, that's really bad.
Another thing I could suggest - Install beta on separate volume.
https://support.apple.com/en-us/HT208891

**Reply by Coeur** [reply]
On the other hand, nevack, that issue was there with Sonoma beta 1-4, Xcode 15 beta 1-5, and after 7 weeks, no one has posted about it on StackOverflow, so there is the possibility that it's still there for the final release? Or it's some particularity in our codebase that makes view rendering different when building with Xcode or building with xcodebuild?

**Reply by Coeur** [reply]
OK, I'm on Sonoma Release Candidate and Xcode 15 RC 1, and Transmission won't work properly when built with Xcode 15.
So @mikedld, please don't adopt Xcode 15 on Sonoma for Transmission builds: we'll need to stay on either Xcode 14 on Sonoma or Xcode 15 on Ventura for making Transmission releases until we figure out what's going on when mixing Sonoma with Xcode 15 in our project.

**Reply by Coeur** [reply]
Finally, I figured it out, after three months with Transmission unusable with Xcode 15:
It's not a bug, it's an undocumented feature change!
#6016 for the fix
